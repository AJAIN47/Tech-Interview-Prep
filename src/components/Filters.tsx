import React, { useState } from 'react';
import { Filter } from 'lucide-react';

// Interface defining the props for the Filters component
interface FiltersProps {
  // Callback function to handle filter changes
  onFilterChange: (filters: FilterState) => void;
}

// Interface defining the filter state structure
interface FilterState {
  // Array of selected question types
  questionTypes: string[];
  // Selected interview duration filter
  interviewDuration: string | null;
}

// Filters component provides filtering options for company list
// Props:
// - onFilterChange: Callback to handle filter changes
const Filters: React.FC<FiltersProps> = ({ onFilterChange }) => {
  // State for filter panel visibility
  const [isOpen, setIsOpen] = useState(false);
  // Current filter state
  const [filters, setFilters] = useState<FilterState>({
    questionTypes: [],
    interviewDuration: null,
  });

  // Available question type filter options
  const questionTypeOptions = [
    'Coding',
    'System Design',
    'Behavioral',
    'Domain Knowledge',
    'Product Sense',
  ];

  const durationOptions = [
    'Under 3 weeks',
    '3-5 weeks',
    '6-8 weeks',
    'Over 8 weeks',
  ];

  // Toggle filter panel visibility
  const toggleFilter = () => {
    setIsOpen(!isOpen);
  };

  // Handle question type filter changes
  // Toggles the selection of a question type and updates filters
  const handleQuestionTypeChange = (type: string) => {
    const updatedTypes = filters.questionTypes.includes(type)
      ? filters.questionTypes.filter(t => t !== type)
      : [...filters.questionTypes, type];
    
    const updatedFilters = {
      ...filters,
      questionTypes: updatedTypes,
    };
    
    setFilters(updatedFilters);
    onFilterChange(updatedFilters);
  };

  // Handle interview duration filter changes
  // Toggles the selection of a duration filter
  const handleDurationChange = (duration: string) => {
    const updatedFilters = {
      ...filters,
      interviewDuration: filters.interviewDuration === duration ? null : duration,
    };
    
    setFilters(updatedFilters);
    onFilterChange(updatedFilters);
  };

  // Clear all active filters
  const clearFilters = () => {
    const resetFilters = {
      questionTypes: [],
      interviewDuration: null,
    };
    
    setFilters(resetFilters);
    onFilterChange(resetFilters);
  };

  return (
    <div className="mb-6">
      <div className="flex items-center justify-between">
        {/* Toggle filter panel button with active count indicator */}
        <button
          onClick={toggleFilter}
          className="flex items-center px-4 py-2 bg-indigo-100 dark:bg-indigo-900 text-indigo-800 dark:text-indigo-200 rounded-lg hover:bg-indigo-200 dark:hover:bg-indigo-800 transition-colors duration-200"
        >
          <Filter className="w-4 h-4 mr-2" />
          <span>Filters</span>
          {/* Active filters count indicator */}
          {(filters.questionTypes.length > 0 || filters.interviewDuration) && (
            <span className="ml-2 inline-flex items-center justify-center w-5 h-5 text-xs font-bold bg-indigo-600 text-white rounded-full">
              {filters.questionTypes.length + (filters.interviewDuration ? 1 : 0)}
            </span>
          )}
        </button>
        
        {/* Clear filters button - only visible when filters are active */}
        {(filters.questionTypes.length > 0 || filters.interviewDuration) && (
          <button
            onClick={clearFilters}
            className="text-sm text-gray-600 dark:text-gray-400 hover:text-gray-900 dark:hover:text-gray-200 transition-colors duration-200"
          >
            Clear all
          </button>
        )}
      </div>
      
      {isOpen && (
        <div className="mt-4 p-4 bg-white dark:bg-gray-800 rounded-lg shadow-lg border border-gray-200 dark:border-gray-700">
          <div className="grid grid-cols-1 md:grid-cols-2 gap-6">
            <div>
              <h3 className="font-medium text-gray-900 dark:text-white mb-3">Question Types</h3>
              <div className="space-y-2">
                {questionTypeOptions.map((type) => (
                  <label key={type} className="flex items-center">
                    <input
                      type="checkbox"
                      className="w-4 h-4 text-indigo-600 rounded focus:ring-indigo-500 border-gray-300 dark:border-gray-600"
                      checked={filters.questionTypes.includes(type)}
                      onChange={() => handleQuestionTypeChange(type)}
                    />
                    <span className="ml-2 text-gray-700 dark:text-gray-300">{type}</span>
                  </label>
                ))}
              </div>
            </div>
            
            <div>
              <h3 className="font-medium text-gray-900 dark:text-white mb-3">Interview Duration</h3>
              <div className="space-y-2">
                {durationOptions.map((duration) => (
                  <label key={duration} className="flex items-center">
                    <input
                      type="radio"
                      name="duration"
                      className="w-4 h-4 text-indigo-600 focus:ring-indigo-500 border-gray-300 dark:border-gray-600"
                      checked={filters.interviewDuration === duration}
                      onChange={() => handleDurationChange(duration)}
                    />
                    <span className="ml-2 text-gray-700 dark:text-gray-300">{duration}</span>
                  </label>
                ))}
              </div>
            </div>
          </div>
        </div>
      )}
    </div>
  );
};

export default Filters;