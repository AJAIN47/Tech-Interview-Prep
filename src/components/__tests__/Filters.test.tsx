import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import Filters from '../Filters';

describe('Filters Component', () => {
  const mockOnFilterChange = jest.fn();


  test('renders filter button and options', () => {
    render(<Filters onFilterChange={mockOnFilterChange} />);
    
    // Check filter button
    const filterButton = screen.getByRole('button', { name: /filters/i });
    expect(filterButton).toBeInTheDocument();
    
    // Check filter options are not visible initially
    expect(screen.queryByText('Question Types')).not.toBeInTheDocument();
    expect(screen.queryByText('Interview Duration')).not.toBeInTheDocument();
  });

  test('handles question type selection', () => {
    render(<Filters onFilterChange={mockOnFilterChange} />);
    
    // Open filters
    const filterButton = screen.getByRole('button', { name: /filters/i });
    fireEvent.click(filterButton);
    
    // Check options are visible
    expect(screen.getByText('Question Types')).toBeInTheDocument();
    expect(screen.getByText('Interview Duration')).toBeInTheDocument();
    
    // Select a question type
    const codingCheckbox = screen.getByLabelText('Coding');
    fireEvent.click(codingCheckbox);
    
    // Verify filter change was called
    expect(mockOnFilterChange).toHaveBeenCalledWith({
      questionTypes: ['Coding'],
      interviewDuration: null
    });
  });

  test('handles interview duration selection', () => {
    render(<Filters onFilterChange={mockOnFilterChange} />);
    
    // Open filters
    fireEvent.click(screen.getByRole('button', { name: /filters/i }));
    
    // Select duration
    fireEvent.click(screen.getByText('3-5 weeks'));
    
    // Verify duration was selected
    expect(mockOnFilterChange).toHaveBeenCalledWith({
      questionTypes: [],
      interviewDuration: '3-5 weeks'
    });
  });

  test('matches snapshot', () => {
    const { container } = render(<Filters onFilterChange={mockOnFilterChange} />);
    expect(container).toMatchSnapshot();
  });
});
