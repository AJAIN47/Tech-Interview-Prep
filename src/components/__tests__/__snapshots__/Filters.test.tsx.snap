// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters Component matches snapshot 1`] = `
<div>
  <div
    class="mb-6"
  >
    <div
      class="flex items-center justify-between"
    >
      <button
        class="flex items-center px-4 py-2 bg-indigo-100 dark:bg-indigo-900 text-indigo-800 dark:text-indigo-200 rounded-lg hover:bg-indigo-200 dark:hover:bg-indigo-800 transition-colors duration-200"
      >
        <svg
          class="lucide lucide-filter w-4 h-4 mr-2"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polygon
            points="22 3 2 3 10 12.46 10 19 14 21 14 12.46 22 3"
          />
        </svg>
        <span>
          Filters
        </span>
      </button>
    </div>
  </div>
</div>
`;
